[WIP] Add type promotion to a fast path for foreach APIs

izdeby opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #52450 wip
* #49039 [wip] Replace optimizers in torch.optim with the ones from torch.optim._multi_tensor
* **#52449 [WIP] Add type promotion to a fast path for foreach APIs**
* #52448 [WIP] Support broadcasting via slow path
* #48224 [wip] Pushed unsupported scenarios to slow path
* #51170 Updated foreach pointwise ops to support complex and bloat16 on CUDA
* #49714 Update foreach max/min
* #51059 Refactor foreach binary ops tests with tensor lists to use OpInfo
* #51058 Refactor foreach binary ops tests with scalars to use OpInfo

Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

Commit:
Add type promotion to a fast path for foreach APIs

[ghstack-poisoned]

Commit:
Update on "Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

Commit:
Update on "Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

Commit:
Update on "Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/ForeachBinaryOpScalarList.cu:
```diff
@@ -6,18 +6,23 @@
+
+    ScalarType result_type = get_result_type(tensors[0], scalars[0], promote_integer_float);
+    for (int i = 0; i < tensors.size(); i++) {
+        tensor_lists[0].emplace_back(tensors[i].to(result_type));
```
lest I forget, this is not a good way to do type promotion - it will launch as many kernels as "slow" path (each `.to` launches a kernel), and will probably result in more memory traffic.

Commit:
Update on "Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

Commit:
Update on "Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

Commit:
Update on "Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

Commit:
Update on "Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

Commit:
Update on "Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

izdeby committed:
Update on "Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

izdeby committed:
Update on "Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

izdeby committed:
Update on "Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

izdeby committed:
Update on "Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

izdeby committed:
Update on "Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

izdeby committed:
Update on "[WIP] Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

izdeby committed:
Update on "[WIP] Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

izdeby committed:
Update on "[WIP] Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]

Commit:
Update on "[WIP] Add type promotion to a fast path for foreach APIs"


Differential Revision: [D26520924](https://our.internmc.facebook.com/intern/diff/D26520924)

[ghstack-poisoned]
